### ISSUE TITLE
Keyboard joystick: expose even if controller connected; 12 button support

### ISSUE THREAD

**Issue by BruceMcF** [commanderx16/x16-rom#168]
I have a previous issue I raised which I closed in favor of this one.

The workaround for finding out whether the Control or Alt keys are being pressed is to just use the JOY Kernel call, IF no joystick is attached.

The proposal here is to use the JOY command with a joystick number of 0 to do this no matter whether or not Joysticks are attached.

This is BOTH for applications like text editors AND for games that include keyboard input, so it is specified to work for both.

REVIEW: If no gamepad is attached, the keyboard can be used as an ersatz gamepad with:
RegA:
NES: [A][B][SEL][START][UP][DN][LT][RT]
SNES: [B][Y][SEL][START][UP][DN][LT][RT]
KEY: [CTRL][ALT][SPACE][ENTER][K-UP][K-DOWN][K-LEFT][K-RIGHT]

RegX:
NES: [0][0][0][0][0][0][0][0] {*1}
SNES: [A][X][TRL][TRR][1][1][1][1]
KEY: [0][0][0][0][0][0][0][1]

RegY:
NES: #0 = Joystick Present, #$FF = Joystick absent
SNES: #0 = Joystick Present, #$FF = Joystick absent

FIRST, the extension: in addition to returning Control, Shift, and cursor keys state (needed for games where the d-pad does different things when A or B are being held down at the same time), return:

RegX:
NES: [0][0][0][0][0][0][0][0] {*1}
SNES: [A][X][TRL][TRR][1][1][1][1]
KEY: [SHIFT][KEY?][LOGO][MENU][0][0][0][1]

I: Gaming Rationale: SHIFT, LOGO (the CX16 8bit catrpiller transformed into a beautiful new butterfly), LOGO and MENU allow all of the SNES keyset to be emulated. EVEN FOR EIGHT BIT CONTROL GAMES this gives flexibility for button remapping to fit what it most comfortable to the player:
I.1] LOGO and MENU are Trigger Left and Trigger Right because (at least on the deluxe keyboard) they are next to each other, and the LOGO key is the one on the left. [MENU][LOGO] would drive people crazy.
I.2] Among "modifier" type keys, that leaves SHIFT. If in 12 button games, CTRL is equivalent to B and Alt is Equivalent to Y, then Shift is "above" Ctrl and ... on the right hand side of the deluxe keyboard ... to the right. So make Shift the SNES A on the right hand side of the diamond.
I.3] So what is "KEY?" It is ANY OTHER KEY is present in the Keyboard Buffer.

II: BEHAVIOR OF KEY? and JOY(0) with the KEYBOARD BUFFER:
II.1] A game that RELIES on keyboard control will want to call JOY(0) to FIND OUT if a modified key is being pressed OR if an unmodified key is being pressed OR if no key is being pressed.
II.2] However, while the modifiers are known SPECIFICALLY, the other keys are only known in terms of SOMETHING IS THERE. So CHRIN (this is key entry, not line entry GETIN) needs to be called to GET it.
II.3] But we want to know whether a NEW key has been pressed when we call JOY(0) again.
II.4] Therefore when there IS a key to be read, it is NOT cleared when JOY(0) is called to report that it is there, so that CHARIN can be called to find out what key it is.
II.5] HOWEVER to use "any other key" as X, it cannot stay "stuck on". EACH call to JOY(1) should report on a NEW key being pressed.
II.6] THEREFORE, when KEY? was 0 in the previous call to JOY, (whether JOY(0) or a missing joystick call to JOY(1)), and there is now a key, the key is NOT cleared. But if KEY? was 1 in the previous call to JOY, and CHARIN was not called in the interim, the keyboard buffer is advanced and the report is whether there is ANOTHER key.
II.7] This is a classical "dirty bit" problem: JOY(0) examines the bit, and only advances the keyboard buffer IF the dirty bit is set, THEN stores the new value of the KEY? bit in the dirty bit. Code is added to CHRIN to clear the dirty bit every time it is called.
II.8] THIS AUTOMATICALLY HANDLES ROLLOVER! You simply keep reading CHARIN until it returns 0 to get all multiple key pressed that the system has recorded. BUT ONLY IF YOUR ARE QUICK: obviously the system does not store THAT these are rollover keys, it just stores as many rolled over keys as it stores, and it's up to the program to clear out the buffer fast enough so that it doesn't confuse successive key events for a rolled over event.
II.9] THIS IS KIND TO SIMPLE GAMES: Games that do not require 12 keys can simply ignore KEY? and never need to call CHARIN.

III. Use in Applications
This is not a multi-part one, because it's already been addressed. Writing a text editor FROM A KEYBOARD EVENT PERSPECTIVE is just like writing a keyboard controlled game, it's just that the result is supposed to be non-surprising and (typically) non-exciting. But with the same JOY(0) system, a WordStar clone with the addition of dedicated cursor and special keys could easily be written, where for a C64 it would requires all sorts of diving down into "magic locations" to get the equivalent.

_________________________________
Notes:
{*1 This is a physical result of the 8 bit shift register being used ... it is a parallel to serial shift register, with the read of the parallel lines triggered by pulsing the SEL line. So once you have read 8 bits, the NES controller will return 0 from then on, no matter how many you read. That is why when Nintendo added the four extra buttons, they used two 8bit serial shift registers connected in series and connected the final four bits of parallel lines to VCC, to "stamp" the ID into the upstream shift register.}

**Comment by mist64** [comment]
I like it.

**Comment by mist64** [comment]
Related: #232

**Comment by mist64** [comment]
Related: #194

**Comment by mist64** [comment]
This is what I decided you do:
* Keyboard modifiers can already be retrieved with `kbd_get_modifiers` now, so there is no need to use the joystick API for this any more.
* Joysticks will be numbers 0-4, 0 is always the keyboard joystick, and 1-4 are the SNES controllers. There is no fallback logic in the KERNAL; games have to worry about this - but it's really trivial, since the API tells you whether a joystick is present at all.
* I went with the SNES9X/ZSNES keyboard layout for mapping A-B-X-Y-L-R to the keyboard, which is a pretty standardized way to map SNES controllers to a keyboard. (This also means that SELECT moved to LShift.)
* I kept the Ctrl -> A, Alt -> B mapping. For simple games that only use A and B, these keys are easier to reach. The left-right order is reversed from the SNES9X/ZSNES mapping of A/B  ('X' and 'Z' keys), which seems to be more in line with DOS gaming (e.g. Commander Keen jumps on Ctrl = A).
